pyuavcan/transport/can/media/pythoncan/_pythoncan.py

```diff
@@ -4,54 +4,343 @@
+                                 bitrate=parameters.bitrate)
+    elif isinstance(parameters, _FDInterfaceParameters):
+        """
+        These magic numbers come from the settings of PCAN adapter.
+        They don't allow any direct baudrate settings, you have to set all lengths and value of the main frequency. 
+        Bit lengths below are very universal and can be applied for almost every popular baudrate. 
+        There is probably a better solution here, but it needs significantly more time to implement it.
+        """
+        f_clock = 40000000
+        nom_tseg1, nom_tseg2, nom_sjw = 3, 1, 1
+        data_tseg1, data_tseg2, data_sjw = 3, 1, 1
+
+        nom_br = int(f_clock / parameters.bitrate[0] / (nom_tseg1 + nom_tseg2 + nom_sjw))
+        data_br = int(f_clock / parameters.bitrate[1] / (data_tseg1 + data_tseg2 + data_sjw))
+        # TODO: validate the result and see if it is within an acceptable range
+        
+        return can.ThreadSafeBus(interface=parameters.interface_name,
+                                 channel=parameters.channel_name,
+                                 f_clock=f_clock,
+                                 nom_brp=nom_br,
+                                 data_brp=data_br,
+                                 nom_tseg1=nom_tseg1,
+                                 nom_tseg2=nom_tseg2,
+                                 nom_sjw=nom_sjw,
+                                 data_tseg1=data_tseg1,
+                                 data_tseg2=data_tseg2,
+                                 data_sjw=data_sjw,
+                                 fd=True)
+    else:
+        raise TypeError(f'Invalid parameters: {parameters}')
+
+
+def _construct_nican(parameters: _InterfaceParameters) -> can.ThreadSafeBus:
+    if isinstance(parameters, _ClassicInterfaceParameters):
+        return can.ThreadSafeBus(interface=parameters.interface_name,
+                                 channel=parameters.channel_name,
+                                 bitrate=parameters.bitrate)
+    elif isinstance(parameters, _FDInterfaceParameters):
+        raise TypeError('Interface does not support CAN FD: {}'.format(parameters.interface_name))
+    else:
+        raise TypeError(f'Invalid parameters: {parameters}')
+
+
+def _construct_any(parameters: _InterfaceParameters) -> can.ThreadSafeBus:
+        raise TypeError('Interface not supported yet: {}'.format(parameters.interface_name))
+
+
+_CONSTRUCTORS: typing.DefaultDict[str, typing.Callable[[_InterfaceParameters],
+                                                       can.ThreadSafeBus]] = collections.defaultdict(
+        lambda: _construct_any, {'socketcan': _construct_socketcan,
+                                 'kvaser':    _construct_kvaser,
+                                 'slcan':     _construct_slcan,
+                                 'pcan':      _construct_pcan,
+                                 'nican':     _construct_nican,}
+    )
+
+
 class PythonCANMedia(_media.Media):
```

Comment:
Can you please move the internals towards the bottom and keep this class near the top? It's not C, no forward declarations are needed here.